build-system/tasks/e2e/mocha-custom-json-reporter.js

```diff
@@ -50,14 +56,14 @@ function JsonReporter(runner) {
   const testEvents = [];
   let suiteList = [];
 
+  // We need a fresh array copy every time we enter a new suite,
+  // so we can't use push or pop here (because the suiteList info of previous
+  // tests in the same suite would be changed)
   runner.on(EVENT_SUITE_BEGIN, function (suite) {
-    suiteList.push(suite.title);
+    suiteList = suiteList.concat([suite.title]);
```

Comment:
The previous method assumed that suites couldn't contain a mixture of suites and tests; the assumption doesn't seem to be correct.

Reply:
Maybe include a counterexample here in the comments (if one is readily available) for posterity?

Reply:
Thinking of the suite structure as a tree, `suitelist.push` there only works if tests are only at leaf nodes. Example:
```js
describe('animals', () => {
  it('breathes', () => { /* test that animal class can breathe */ });
  describe('felines', () => {
    it('has cute cubs' () => { /* test cub cuteness */ });
  });
};
```
With the push method, we would get the following:
`['animals', 'felines', 'breathes'], ['animals', 'felines', 'has cute cubs']`